### PR TITLE
BAH-4680 | [Registration] Migrate Patient Profile Fetch to FHIR R4 API

### PR DESCRIPTION
**Pull request by Varun-beta**
## Summary

Migrates patient profile loading on the edit page from OpenMRS REST API (`/patientprofile/{uuid}?v=full`) to FHIR R4 API (`GET /fhir2/R4/Patient/{uuid}`).

### What's changed

- **`usePatientDetails`** — now calls `getPatientById` (FHIR GET) instead of `getPatientProfile` (REST GET)
- **`fhirPatientToFormData.ts`** (new) — reverse mapper: FHIR Patient → form data (profile, address, attributes, identifiers, metadata)
- **`fhirUtils.ts`** (new) — shared constants and helpers (`toSlugCase`, `mapGenderToFhir`, `mapGenderFromFhir`, FHIR extension URLs)
- **`fhirPatientMapper.ts`** — refactored to import from `fhirUtils.ts` instead of defining its own helpers
- **Relationships** — removed from patient details flow (out of scope, not in FHIR Patient resource)

### Field mapping (FHIR → Form)

| Form field | FHIR source |
|---|---|
| Name | `name[0].given` / `family` |
| Gender | `gender` → reverse map (male → M) |
| Date of birth | `birthDate` (YYYY = estimated, YYYY-MM-DD = exact) |
| Birth time | `_birthDate.extension[patient-birthTime]` |
| Estimated flag | Inferred from `birthDate` precision |
| Person attributes | `extension[]` with `ext/patient/{slug}` → slug reversed to attribute name |
| Address | `address[0]` fields + `ext/address#address1/address2` extensions |
| Identifiers | `identifier[]` — primary (index 0) + additional |
| Registration date | `extension[ext/patient/date-created]` (custom, from backend override) |

### Depends on

- Backend: `BahmniPatientTranslatorImpl` in `bahmni-module-fhir2-addl-extension` — provides person attribute extensions, birth time, and `date-created` extension on GET

### Files

| File | Change |
|---|---|
| `fhirUtils.ts` | New — shared FHIR constants + helpers |
| `fhirPatientToFormData.ts` | New — FHIR Patient → form data converters |
| `fhirPatientMapper.ts` | Refactored imports from `fhirUtils.ts` |
| `usePatientDetails.ts` | FHIR GET + new converters |
| `PatientRegister.tsx` | Removed `relationshipsInitialData` |
| `formSectionMap.tsx` | Removed `initialData` from PatientRelationships |
| `models.ts` | Removed `relationshipsInitialData` field |

## Test plan

- [ ] Edit patient — all fields populate correctly from FHIR GET
- [ ] Name, gender, DOB, birth time, address, phone, email — all render
- [ ] Estimated birthdate — checkbox checked, age calculated from YYYY
- [ ] Additional identifiers shown correctly
- [ ] Registration date displayed from `date-created` extension
- [ ] Error notification shown on fetch failure
- [ ] No fetch when patientUuid is undefined

## Summary by CodeRabbit

* **Refactor**
  * Patient data loading and form-initialization redesigned to consume FHIR patient records and shared FHIR helpers.
* **New Features**
  * More robust extraction of names, gender, DOB (including estimated DOB handling), contact attributes, addresses, and extra identifiers from FHIR data.
* **Bug Fixes**
  * Hook now surfaces fetch errors as notifications and avoids fetching when no patient ID is provided.
* **Tests**
  * Expanded test coverage for FHIR-to-form conversions and gender/slug utilities.

### CODE REVIEW

File: apps/registration/src/utils/fhirUtils.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
+export const PATIENT_ATTRIBUTE_PREFIX = 'http://fhir.bahmni.org/ext/patient/'; // NOSONAR
+export const ADDRESS_EXT_URL = 'http://fhir.openmrs.org/ext/address'; // NOSONAR
+export const BIRTH_TIME_EXT_URL =
+  'http://hl7.org/fhir/StructureDefinition/patient-birthTime'; // NOSONAR
+export const IDENTIFIER_LOCATION_EXT_URL =
+  'http://fhir.openmrs.org/ext/patient/identifier#location'; // NOSONAR
+export const DATE_CREATED_EXT_URL = PATIENT_ATTRIBUTE_PREFIX + 'date-created';
+
+export function toSlugCase(str: string): string {
```
Comment (ravinderkabli) [inline apps/registration/src/utils/fhirUtils.ts:9]: nit: This file mixes URL constants (lines 1-7) and utility functions (`toSlugCase`, `mapGenderToFhir`, `mapGenderFromFhir`). `DATE_CREATED_EXT_URL` on line 7 is also a Bahmni-specific domain constant only consumed by `extractMetadata` in `fhirPatientToFormData.ts`.

Consider splitting constants into `fhirConstants.ts` and keeping helpers here, or at minimum moving `DATE_CREATED_EXT_URL` to `fhirPatientToFormData.ts` where it's the sole consumer.

File: apps/registration/src/utils/__tests__/fhirPatientToFormData.test.ts [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+
+describe('extractDobEstimated', () => {
+  it('should return true for YYYY', () => {
+    expect(extractDobEstimated({ ...baseFhirPatient, birthDate: '1990' })).toBe(
+      true,
+    );
+  });
+
+  it('should return true for YYYY-MM', () => {
+    expect(
```
Comment (ravinderkabli) [inline apps/registration/src/utils/__tests__/fhirPatientToFormData.test.ts:269]: nit: `extractMetadata` has only one test (happy path). Three edge-case code paths are untested:

1. Patient with no `date-created` extension → `registerDate` should be `''`
2. Patient with no `identifier` → `patientIdentifier` should be `''`
3. Patient with no `name` → `patientName` should be `''`

The function has optional chaining for all three paths but none are exercised. Consider adding test cases for these.

**Review by ravinderkabli — COMMENTED** [review]
## Review for BAH-4680

Clean migration from REST to FHIR R4 for patient profile fetching. Well-structured reverse mapper with symmetric read/write mapping and thorough test coverage.

File: apps/registration/src/utils/fhirUtils.ts (hunk 1 above)
Comment (Varun-beta) [reply apps/registration/src/utils/fhirUtils.ts:9]: Fixed. Moved `DATE_CREATED_EXT_URL` to `fhirUtils.ts` as a centralized constant (consistent with the backend change in PR #78 to move it to `BahmniFhirConstants`). Also updated the URL to `/patient-record/date-created` to avoid the person attribute prefix collision.

File: apps/registration/src/utils/__tests__/fhirPatientToFormData.test.ts (hunk 2 above)
Comment (Varun-beta) [reply apps/registration/src/utils/__tests__/fhirPatientToFormData.test.ts:269]: Fixed. Added 3 edge-case tests for `extractMetadata`: no `date-created` extension, no `identifier`, and no `name`.

File: apps/registration/src/utils/fhirPatientToFormData.ts [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+function extractBirthTime(patient: Patient): string {
+  const el = (patient as Record<string, unknown>)._birthDate as
+    | { extension?: { url: string; valueDateTime?: string }[] }
+    | undefined;
+  const ext = el?.extension?.find((e) => e.url === BIRTH_TIME_EXT_URL);
+  if (!ext?.valueDateTime) return '';
+  const date = parseISO(ext.valueDateTime);
+  return isValid(date) ? format(date, 'HH:mm') : '';
+}
```
[the remark this answers is not in the document]
Comment (Varun-beta) [reply apps/registration/src/utils/fhirPatientToFormData.ts:38]: Not a bug. The backend stores birth time in UTC (`+00:00`), and `format(date, 'HH:mm')` from date-fns correctly renders it in the user's local timezone — which is the standard and expected behavior. No change needed.

**Review by ravinderkabli — APPROVED** [review]
Looks fine
